src/storage/db.rs

```diff
@@ -259,3 +261,58 @@ impl super::HashInterface for Storage {
         }
     }
 }
+
+#[async_trait::async_trait]
+impl super::FingerprintInterface for Storage {
+    type Error = error::FingerprintDBError;
+
+    async fn find_by_card_hash(
+        &self,
+        card_hash: &[u8],
+    ) -> Result<Option<types::Fingerprint>, ContainerError<Self::Error>> {
+        let mut conn = self.get_conn().await?;
+
+        let output: Result<_, diesel::result::Error> = types::Fingerprint::table()
+            .filter(schema::fingerprint::card_hash.eq(card_hash))
+            .get_result(&mut conn)
+            .await;
+
+        match output {
+            Ok(inner) => Ok(Some(inner)),
+            Err(inner_err) => match inner_err {
+                diesel::result::Error::NotFound => Ok(None),
+                error => Err(error).change_error(error::StorageError::FindError)?,
+            },
+        }
+    }
+    async fn insert_fingerprint(
+        &self,
+        card: types::CardNumber,
+        hash_key: Secret<String>,
+    ) -> Result<types::Fingerprint, ContainerError<Self::Error>> {
+        let algo = HmacSha512::<1>::new(hash_key.expose().into_bytes().into());
+        let card_data = serde_json::to_vec(&card)
+            .change_error(error::FingerprintDBError::EncodingError("byte-vec"))?;
+
+        let card_hash = algo.encode(card_data)?;
```

Comment:
I feel the crypto error variant can be handled properly to give more error insights, as mentioned in one of the above comment